bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/AbstractFileProvider.java

```diff
@@ -0,0 +1,340 @@
+                }
+                value.add(url);
+            }
+            logger.debug("Parser {} not available", parserType, new Exception());
+        }
+
+    }
+
+    @Override
+    protected AbstractWatchQueueReader buildWatchQueueReader(WatchService watchService, Path toWatch) {
+        return new WatchQueueReader(watchService, toWatch, this);
+    }
+
+    @Override
+    protected boolean watchSubDirectories() {
+        return true;
+    }
+
+    @Override
+    protected void registerDirectory(Path subDir) throws IOException {
+        subDir.register(watchService, ENTRY_CREATE, ENTRY_DELETE, ENTRY_MODIFY);
+    }
+
+    protected void notifyListeners(E oldElement, E newElement) {
+        synchronized (listeners) {
+            for (ProviderChangeListener<E> listener : listeners) {
+                if (oldElement != null) {
+                    listener.updated(this, oldElement, newElement);
+                }
+                listener.added(this, newElement);
+            }
+        }
+    }
+
+    protected void notifyListeners(E removededObject) {
+        if (removededObject != null) {
+            synchronized (listeners) {
+                for (ProviderChangeListener<E> listener : listeners) {
+                    listener.removed(this, removededObject);
+                }
+            }
+        }
+    }
+
+    protected E getOldElement(String uid) {
+        synchronized (providedObjectsHolder) {
+            return providedObjectsHolder.get(uid);
+        }
+    }
+
+    protected void updateProvidedObjectsHolder(URL url, Set<E> providedObjects) {
+        if (providedObjects != null && !providedObjects.isEmpty()) {
+            List<String> uids = new ArrayList<String>();
+            for (E providedObject : providedObjects) {
+                String uid = getUID(providedObject);
+                E oldModuleType = getOldElement(uid);
+                notifyListeners(oldModuleType, providedObject);
+                uids.add(uid);
+                synchronized (providedObjectsHolder) {
+                    providedObjectsHolder.put(uid, providedObject);
+                }
+            }
+            synchronized (providerPortfolio) {
+                providerPortfolio.put(url, uids);
+            }
+        }
+    }
+
+    protected abstract String getUID(E providedObject);
+
+    protected void removeElements(List<String> objectsForRemove) {
+        if (objectsForRemove != null) {
+            for (String removededObject : objectsForRemove) {
+                E mtRemoved;
+                synchronized (providedObjectsHolder) {
+                    mtRemoved = providedObjectsHolder.remove(removededObject);
+                }
+                notifyListeners(mtRemoved);
+            }
+        }
+    }
+
+    private String getParserType(URL url) {
+        String fileName = url.getPath();
+        int fileExtesionStartIndex = fileName.lastIndexOf(".") + 1;
+        if (fileExtesionStartIndex == -1) {
```

Comment:
fileExtesionStartIndex will never be -1 even when there is not an file extension.